engine/consensus/compliance/engine.go

```diff
@@ -149,9 +173,108 @@ func (e *Engine) ProcessLocal(event interface{}) error {
+			return val.(*Event), e.blockSink, e.pendingBlocks
+		}
+		if val, ok := e.pendingVotes.Front(); ok {
+			return val.(*Event), e.voteSink, e.pendingVotes
+		}
+		return nil, nil, nil
+	}
+
+	for {
+		pendingEvent, sink, fifo := fetchEvent()
+		select {
+		case event := <-e.pendingEventSink:
+			e.processPendingEvent(event)
+		case sink <- pendingEvent:
+			fifo.Pop()
+			continue
+		case <-e.unit.Quit():
+			return
+		}
+	}
+}
+
+// processPendingEvent saves pending event in corresponding queue for further processing by `Core`.
+// While this function runs in separate goroutine it shouldn't do heavy processing to maintain efficient data polling/pushing.
+func (e *Engine) processPendingEvent(event *Event) {
+	// skip any message as long as we don't have the dependencies
+	if e.core.hotstuff == nil || e.core.sync == nil {
+		return
+	}
+
+	switch event.Msg.(type) {
+	case *events.SyncedBlock:
+		e.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageSyncedBlock)
+		// a block that is synced has to come locally, from the synchronization engine
+		// the block itself will contain the proposer to indicate who created it
+		if event.OriginID != e.me.NodeID() {
+			log.Warn().Msgf("synced block with non-local origin (local: %x, origin: %x)", e.me.NodeID(), event.OriginID)
+			return
+		}
+		e.pendingBlocks.Push(event)
+	case *messages.BlockProposal:
+		e.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageBlockProposal)
+		e.pendingBlocks.Push(event)
+	case *messages.BlockVote:
+		e.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageBlockVote)
+		e.pendingVotes.Push(event)
+	}
+}
+
+// consumeEvents consumes events that are ready to be processed.
+func (e *Engine) consumeEvents() {
+	processBlock := func(event *Event) error {
+		var err error
+		switch t := event.Msg.(type) {
+		case *events.SyncedBlock:
+			proposal := &messages.BlockProposal{
+				Header:  t.Block.Header,
+				Payload: t.Block.Payload,
+			}
+			err = e.core.OnBlockProposal(event.OriginID, proposal)
+			e.metrics.MessageHandled(metrics.EngineCompliance, metrics.MessageSyncedBlock)
+
+		case *messages.BlockProposal:
+			err = e.core.OnBlockProposal(event.OriginID, t)
+			e.metrics.MessageHandled(metrics.EngineCompliance, metrics.MessageBlockProposal)
+		}
+		return err
+	}
+
+	for {
+		var err error
+		select {
+		case event := <-e.blockSink:
+			err = processBlock(event)
+		case event := <-e.voteSink:
+			err = e.core.OnBlockVote(event.OriginID, event.Msg.(*messages.BlockVote))
+			e.metrics.MessageHandled(metrics.EngineCompliance, metrics.MessageBlockVote)
+		case <-e.unit.Quit():
+			return
+		}
+		if err != nil {
+			// Public methods of `Core` are supposed to handle all errors internally.
+			// Here if error happens it means that internal state is corrupted or we have caught
+			// exception while processing. In such case best just to abort the node.
+			e.log.Fatal().Err(err).Msgf("fatal internal error in matching core logic")
```

Comment:
I proposed the error handling in [this PR](https://github.com/onflow/flow-go/pull/419/files#diff-17e93dc346462f66b95ec3f31134583765b8fdc4eff5e3c03f276698a554997fL257), better to give a look first and so if we could make the error handling consistent.

Since we will `Fatal` on any error returned by `processBlock` and `OnBlockVote`, could you test on localnet, and see if the consensus node will crash due to some network connectivity failure, which shouldn't. 

For instance, you can test turning off some other consensus nodes, and see if other consensus nodes will crash, which shouldn't. Could also do the same to other node type.